polar-core/src/kb.rs

```diff
@@ -75,11 +81,247 @@ impl KnowledgeBase {
         self.rules.insert(rule.name.clone(), rule);
     }
 
+    pub fn add_rule(&mut self, rule: Rule) {
+        let name = rule.name.clone();
+        let generic_rule = self
+            .rules
+            .entry(name.clone())
+            .or_insert_with(|| GenericRule::new(name, vec![]));
+        generic_rule.add_rule(Arc::new(rule));
+    }
+
+    pub fn validate_rules(&self) -> PolarResult<()> {
+        for (symbol, generic_rule) in self.rules.clone() {
+            let generic_template = self.rule_templates.get(&symbol);
+            if let Some(generic_template) = generic_template {
+                // If a template with the same name exists, then the parameters must match for each rule
+                for (_, r) in generic_rule.rules {
+                    // get templates with correct arity
+                    let templates = generic_template
+                        .rules
+                        .values()
+                        .filter(|t| return t.params.len() == r.params.len())
+                        .collect::<Vec<&Arc<Rule>>>();
+
+                    if templates.is_empty() {
+                        let err: PolarError = error::RuntimeError::FileLoading {
+                            msg: format!("Rule {}({:#?}) not allowed because arity did not match rule template.", r.name, r.params),
+                        }
+                        .into();
+                        return Err(err.set_context(None, Some(&r.body)));
+                    } else {
+                        // these templates should all have matching arity
+                        // TODO: check params of rule against rule templates
+                        let mut found_match = false;
+                        for t in templates {
+                            if self.rule_params_match(r.as_ref(), t.as_ref()) {
+                                found_match = true;
+                                break;
+                            }
+                        }
+                        if !found_match {
+                            let err: PolarError = error::RuntimeError::FileLoading {
+                                msg: format!("Rule {}({:#?}) not allowed because parameters did not match rule template.", r.name, r.params),
+                            }
+                            .into();
+                            return Err(err.set_context(None, Some(&r.body)));
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn param_fields_match(&self, template_fields: &Dictionary, rule_fields: &Dictionary) -> bool {
+        return template_fields
+            .fields
+            .iter()
+            .map(|(k, template_value)| {
+                rule_fields
+                    .fields
+                    .get(k)
+                    .map(|rule_value| rule_value == template_value)
```

Comment:
Does this need to be recursive? What about something like:

```
type foo(a: Foo{a: Bar{b: 1}});

foo(a: Foo{a: Bar{b: 1, c: 2}, d: 2})
```

Reply:
This is related to the bug that prevents pattern specializers from working on fields. Since that doesn't work right now, I didn't account for it here.

Reply:
Ok, maybe leave a todo here?